### PR TITLE
[ECC chip] Incomplete and complete addition

### PR DESCRIPTION
**Pull request by therealyingtong**
The variable- and fixed-base scalar multiplication instructions are implemented in #111.

### CODE REVIEW

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip.rs:143]: Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r640934467

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip.rs:246]: Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r640939832

File: src/circuit/gadget/ecc/chip/add.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip/add.rs:90]: Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r642422385

File: src/circuit/gadget/ecc/chip/add.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip/add.rs:194]: Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r645092459

File: src/circuit/gadget/ecc/chip/add.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip/add.rs:268]: Unresolved comment: https://github.com/zcash/orchard/pull/73#discussion_r645096894

File: src/circuit/gadget/ecc/chip/mul.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip/mul.rs:93]: TODO: add `t_q` (and corresponding `t_p` for Pallas) as constants.

**Comment by str4d** [comment]
Rebased on `main` after #89 and #109 were merged. I squashed the `halo2` API changes into the relevant commits, so we can still review commit-by-commit.

File: src/circuit/gadget/ecc/chip/mul.rs
Comment (ebfull) [inline src/circuit/gadget/ecc/chip/mul.rs:121]: What gate is enforcing that this must be zero?

File: src/circuit/gadget/ecc/chip/mul_fixed/short.rs
Comment (ebfull) [inline src/circuit/gadget/ecc/chip/mul_fixed/short.rs:34]: Another example.

File: src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs
Comment (ebfull) [inline src/circuit/gadget/ecc/chip/mul_fixed/full_width.rs:19]: Another example.

**Review by ebfull — CHANGES_REQUESTED** [review]

File: src/circuit/gadget/ecc/chip/mul.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/mul.rs]: I've added a gate checking this in ae62bb2672c3dd466a5d59da3264afa7a522cfb2. Additionally, we are already checking that:
1. The scalar is decomposed bitwise; and
2. We recover the original scalar using the final value of the running sum.

**Review by str4d — COMMENTED** [review]
Partial review.

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [reply src/circuit/gadget/ecc/chip.rs:27]: I agree, but IMHO there should be accessors. We are not trying to hide that this is an affine short-Weierstrass point representation with (0, 0) as the zero point. (The circuit depends heavily on that.) We're only trying to avoid construction of invalid point representations.

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:43]:
```suggestion
    }

    pub fn x(&self) -> CellValue<C::Base> {
        self.x
    }

    pub fn y(&self) -> CellValue<C::Base> {
        self.y
    }
```

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:167]:
```suggestion
        // Create variable-base scalar mul gate
```

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:173]:
```suggestion
        // Create witness scalar_fixed gate that applies to both full-width and
```

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:180]:
```suggestion
        // Create witness scalar_fixed gate that only applies to short scalars
```

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:186]:
```suggestion
        // Create fixed-base scalar mul gate that is used in both full-width
```

File: src/circuit/gadget/ecc/chip.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip.rs:193]:
```suggestion
        // Create gate that is only used in short fixed-base scalar mul.
```

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip.rs:27]: Should these getters be gadget-level?

**Comment by nikkolasg** [comment]
Hi there - thx for this great work !

One thing I'm wondering is when looking at the API, it seems to only offer two choices:
1. Use a `Fp` as a scalar and then we can multiply with any `Point`. Unfortunately as the comments say, it's not the case that any `Fq` can fit into a `Fp` (if we want to use `Fq` as inputs for example)
2. Use a `Fq` as a scalar (short or not) but then we can only multiply with fixed base pre-defined constants related to Orchard (`OrchardFixedBasesFull<C>`).

Am I reading the API correctly? If so, is there a future where the API could offer to use `Fq` and multiply with any `Point` ? This would be, I believe, quite useful for general consumption outside of the Orchard context. Or is there some fundamentals limitations there ?

Thank you !

**Comment by str4d** [comment]
There isn't a fundamental limitation; it would just need to be a set of instructions that operates on a scalar split across two base field elements. We have no need for that in Orchard so it is not on our critical path, but it would be a nice subsequent contribution.

File: src/circuit/gadget/ecc/chip/mul/incomplete.rs
Comment (therealyingtong) [inline src/circuit/gadget/ecc/chip/mul/incomplete.rs:74]:
```suggestion
        meta.create_gate("Incomplete addition range of variable-base scalar mul", |meta| {
```

**Comment by therealyingtong** [comment]
[Force-push](https://github.com/zcash/orchard/compare/e5e5067c19a13975b654944abcf47a1d3176227c..295e0a251418ee8fb4435fed651632d566ed463a) to split out fixed- and variable-base scalar multiplication into #111.

File: src/circuit/gadget/ecc/chip/mul.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/mul.rs]: Resolving this because the running sum initialisation gate has been implemented at https://github.com/zcash/orchard/pull/111/files#diff-8c85db55a23b99719dabdaf6f8ea93578fe5921af5fc3e3802985a8394fa4fb1R85.

File: src/circuit/gadget/ecc/chip/witness_scalar_fixed/short.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/witness_scalar_fixed/short.rs]: This comment has been moved to #111.

File: src/circuit/gadget/ecc/chip/mul.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/mul.rs]: This comment has been moved to #111.

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip.rs]: This comment has been moved to https://github.com/zcash/orchard/pull/111/files#r649392422

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip.rs]: These documentation fixes have been addressed in c98b15239e82142f3a016da1d0495b1cd703fa24 on #111.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/add.rs]: Resolving this since the MockProver improvements to address this have been merged upstream (https://github.com/zcash/halo2/pull/285, https://github.com/zcash/halo2/pull/286)

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip.rs]: This comment has been moved to https://github.com/zcash/orchard/pull/111/files#r649398154

File: src/circuit/gadget/ecc/chip/witness_point.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/witness_point.rs:57]: I'm in favour of allowing witnessing the identity. Even if we disallow it, the identity could be constructed anyway using complete addition `P + (-P)`. (In fact, this is what the tests are currently doing.)

File: src/circuit/gadget/ecc/chip/add_incomplete.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/add_incomplete.rs]: Removed.

File: src/circuit/gadget/ecc/chip.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip.rs:43]: Should these getters be gadget-level? We currently have `extract_p()` to get the x-coordinate at the gadget-level.

Within the chip, each submodule can access x and y anyway.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/add.rs]: Batch inversion for `alpha`, `beta`, `gamma`, `delta` implemented in https://github.com/zcash/orchard/pull/107/commits/4a0f7bad4e3fc94ddc94d0bf92f6b52b257d2840.

File: src/circuit/gadget/ecc/chip/witness_point.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/witness_point.rs]: Changed to allow witnessing identity in https://github.com/zcash/orchard/pull/107/commits/e86f9f92292ade59bb31976493e28ac8094644b0.

File: src/circuit/gadget/ecc/chip/add_incomplete.rs
Comment (therealyingtong) [reply src/circuit/gadget/ecc/chip/add_incomplete.rs:59]: PR created at https://github.com/zcash/halo2/pull/292

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:156]: This is identical to `secant_line` above, and could be factored out. Similarly for the other sub-components; we could define a set of common sub-components, and then clone them where necessary.

(If this conflicts with, or reverses, an earlier review comment, let me know.)

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:247]: After the above change, this can be a structural match:
```suggestion
            if let Some([alpha, beta, gamma, delta]) = inverses {
                (Some(alpha), Some(beta), Some(gamma), Some(delta))
```

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:291]:
```suggestion
                if x_q == x_p {
                    delta.ok_or(Error::SynthesisError)
                } else {
                    C::Base::zero()
                }
```

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:311]:
```suggestion
                            let three_x_p_sq = C::Base::from_u64(3) * x_p.square();
```

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:346]:
```suggestion
                        lambda.square() - x_p - x_q
```

**Review by str4d — COMMENTED** [review]
I have verified that the complete and incomplete addition gates match the documented constraints in https://github.com/zcash/orchard/pull/54.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:412]: This isn't what I was thinking of when I said "constrain": I mean that we add a circuit constraint that `sum == zero`. This is probably fine for testing purposes, but I worry that with the `Point` gadget implementing `PartialEq`, someone else might write code like `if sum == zero { constrain something }` which results in a non-deterministic circuit.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:429]: Here, we are allocating both the sum and the expected result in the circuit, and then checking equality outside the circuit.

**Review by str4d — COMMENTED** [review]
I'm uncertain about the `PartialEq` impl on `Point`; I think we should have an `EccInstructions::constrain_equal` instruction instead (and an equivalent method on `Point`). But I'll defer to others on this, since AFAIK we don't actually need this instruction in the Orchard circuit.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [inline src/circuit/gadget/ecc/chip/add.rs:418]:
```suggestion
            result.constrain_equal(layouter.namespace(|| "𝒪 + 𝒪 = 𝒪"), zero)?;
```

**Review by str4d — APPROVED** [review]
utACK with a few last requests.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip/add.rs:186]: I'm relying on the fact that I checked these in a previous review.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip/add.rs:109]: These will be indexed from 0 in error messages. Don't worry about this if the invention is to give them labels.

File: src/circuit/gadget/ecc/chip/add_incomplete.rs
Comment (daira) [inline src/circuit/gadget/ecc/chip/add_incomplete.rs:52]: Same comment as above: gate polys are zero-indexed, but it's fine if these are to be labelled later.

**Review by daira — APPROVED** [review]
utACK with non-blocking comments.

File: src/circuit/gadget/ecc/chip/add.rs
Comment (str4d) [reply src/circuit/gadget/ecc/chip/add.rs:109]: Yep, we can go through in a subsequent PR to add labels / make cleanups.

**Comment by str4d** [comment]
Test circuit layout LGTM.

ecc-add-chip-layout

**Review by ebfull — APPROVED** [review]

**Pull request merged**
